### ISSUE TITLE
question about decoder_nll loss

### ISSUE THREAD

**Issue by BaohaoLiao** [madaan/minimal-text-diffusion#7]
Hi, 

thank you again for your clean code. Here I have a question about the decoder_nll loss. 

According to your code, you calculate the decoder_nll loss in this way:
```
x_start_mean = model.model.module.get_embeds(input_ids)
x_start = self.get_x_start(x_start_mean, std)
decoder_nll = self.token_discrete_loss(x_start, get_logits, input_ids)

def token_discrete_loss(self, x_t, get_logits, input_ids):
    logits = get_logits(x_t)  # bsz, seqlen, vocab
    loss_fct = th.nn.CrossEntropyLoss(reduction="none")
    decoder_nll = loss_fct(logits.view(-1, logits.size(-1)), input_ids.view(-1)).view(
        input_ids.shape
    )
    decoder_nll = decoder_nll.mean(dim=-1)
    return decoder_nll
```
which means that x_start is the noisy word embedding. You calculate the cross entropy between the noisy word embedding and input_ids. However, in the [diffusion_lm](https://arxiv.org/pdf/2205.14217.pdf), "We now describe
the inverse process of rounding a predicted x0 back to discrete text" (second sentence in Section 4.2). It seems they use the predicted x_start rather than the original x_start. 

I know the original code also implements it in this way, but it confuses me. Why don't we replace x_start in the function self.token_discrete_loss(x_start, get_logits, input_ids) with model_out_x_start? The noisy word embedding x_start should be very close to the original word embedding, since at the beginning we only add little noise. We don't need to calculate its loss. We should make sure the predicted x_start (model_out_x_start) to be close to the word embedding.

**Comment by BaohaoLiao** [comment]
Solved!

**Comment by kaustubholpadkar** [comment]
Can you please explain?! I have the same question.

**Comment by BaohaoLiao** [comment]
@kaustubholpadkar You can find my discussion about this at https://github.com/Yuanhy1997/SeqDiffuSeq/issues/6

**Comment by kaustubholpadkar** [comment]
Thanks! @BaohaoLiao
